### PR TITLE
Function Login, Register, Get Profile, Edit

### PR DESCRIPTION
**Pull request by Hedo21**
# Description

I have created login, register, get profile, edit functions. The get profile function is still waiting for confirmation on what can be displayed. The edit function is also still waiting for confirmation of what data can be edited.

# JSON Contract :
## Request
1.  Register
request parameter:
```JSON
{ 
   "name" : "string",
   "username" : "string" 
   "email": "string"
   "password": "string" 
}
```
2.  Login
request parameter:
```JSON
{ 
   "email" : "string",
   "password" : "string" 
}
```

## Response
1. Register Success
```JSON
{  
    "status": "success",
    "access_token": "48|TPQ3PJATgblaym6Axo0PdHajd9BjMEQRVxplsxtj",
    "message": "registration success"
}
```

2. Register Failed 

```JSON
{
    "status" : "failed",
    "message" : "name is empty",
}
```

error messages that may appear
```JSON
["name is empty", "username is empty", "email is empty", "password is empty", "The username must not be greater than 10 characters.", "The username has already been taken.", "The email has already been taken.",  "The email must be a valid email address.", "The password must be at least 8 characters."]
```

3. Login Success
```JSON
{
    "status": "success",
    "message": "login success",
    "access_token": "49|rz1dcJqTOGU0rLgLvy8bTYYdOcdZ29rw0qlGugjG",
    "token_type": "bearer"
}
```

4. Login failed
```JSON
{
    "status" : "failed",
    "message" : "email is empty",
}
```
error messages that may appear
```JSON
["email is empty", "password is empty", "wrong email or unregistered email","incorrect password"]
```

## Task reference

My Task reference is from google spreadsheet

[Google Spreadsheet](https://docs.google.com/spreadsheets/d/1GcXMjt723xfKeueJbsqyAHba9llaMJVWozkekc_9-tY/edit#gid=995068209)

## Type of change

Please make any relevant options for the progress of your changes.

- [x] Make login form
- [x] Make register form
- [x] Make form validation for username
- [x] Make form validation for email
- [x] Make form validation for password
- [x] Make form validation for empty form

# How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] User Interface Local
- [x] Postman Response
- [ ] User Test

# Testing screenshot

Please provide us with your test results. This may be a positive or negative test for an API response. Also, provide us with the interface before and after your changes are applied if any. Provide a relevant title.
## Registration Success
   image
## Registration Failed
  1. name is empty
      image
  2. username is empty
      image
  3. email is empty
      image
  4. password is empty
      image
  5. username already taken
      image
  6. username character is more than 10 character 
      image
  7. email is not valid
      image
  8. password character is less than 8 character
      image

## Login Success
   image
## Login Failed
  1. email is empty
      image
  2. password is empty
      image
  3. wrong email or unregistered email
      image
  4. incorrect password
      image

# Checklist:

- [x] My changes do not affect any errors in existing features
- [x] Perform `git pull --rebase origin main` before ready to merge
- [x] No conflict in my changes
- [x] Any conflicts are resolved
- [x] I've follows [the code style guidelines](https://)
- [x] I've follows [the code documentation guidelines](https://)
- [x] I've added and run a test scenario that proves my changes are works
- [x] I've provided  all relevant screenshots
- [x] My changes generate no new warnings and no error
- [ ] Approve by at least  2 or more tech member
- [x]  Approve by the project owner
- [x] Using all relevant labels

### CODE REVIEW

**Review by chamisfum — COMMENTED** [review]
@darmainchiro iki haruse nang folder `heroguru-api-app` kan?

**Review by chamisfum — COMMENTED** [review]
mas @Hedo21 Aku minta tolong nanti response nya kaya gini ya

1. Success
```json
{ 
      "status":"success"
      "access_token": "48|TPQ3PJATgblaym6Axo0PdHajd9BjMEQRVxplsxtj",
      "message": "registration successful"
}
```

2. Register Failed
```json
{
    "status":"failed",
    "message":"invalid email address"
}
```

Tolong tambahkan list message yang mungkin terjadi yaa:

```json
["invalid email address", *****]
```

3. Login Success
```JSON
{
    "status":"success"
    "message": "login success",
    "access_token": "49|rz1dcJqTOGU0rLgLvy8bTYYdOcdZ29rw0qlGugjG",
    "token_type": "Bearer"
    
}
```

4. Login failed
```JSON
{
    "status":"failed"
    "message": "invalid email address"
}
```
Tolong tambahkan list message yang mungkin terjadi yaa:

```json
["email required", "password required", "invalid email", "invalid password", *****]
```

**Comment by Hedo21** [comment]
iya mas kubenerin dulu

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                'name'     => 'required|string|max:255',
+                'username' => 'required|string|max:10|unique:users',
+                'email'    => 'required|string|email|max:255|unique:users',
+                'password' => 'required|string|min:8',
+            ],
+            [
+                'name.required'     => 'Enter Your Name',
+                'username.required' => 'Enter Your Username',
+                'email.required'    => 'Enter Your Email',
+                'password.required' => 'Enter Your Password',
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]: Saran untuk json value pake lowercase saja yaa
```suggestion
                'name.required'     => 'name is empty',
                'username.required' => 'username is empty',
                'email.required'    => 'email is empty',
                'password.required' => 'password is empty',
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            ]
+        );
+        if ($validator->fails()) {
+            return response()->json($validator->errors());
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]: Ini kurang statusnya mas. Nanti jangan lupa di update di deskripsi juga ya hasil testingnya 💪 

```suggestion
return response()->json(
     [
          'status' => 'failed',
          'message' => $validator->errors()->all()
     ]
);
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                'username' => $request->username,
+                'email'    => $request->email,
+                'password' => Hash::make($request->password),
+            ]);
+            if ($user) {
+                $token = $user->createToken('auth_token')->plainTextToken;
+                return response()->json([
+                    'Status'       => 'Success',
+                    'access_token' => $token,
+                    'message'      => 'Registration Success',
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                    'status'       => 'success',
                    'access_token' => $token,
                    'message'      => 'registration success',
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                ], 200);
+            } else {
+                return response()->json([
+                    'status'  => 'Failed',
+                    'message' => 'Registration Failed',
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                    'status'  => 'failed',
                    'message' => 'registration railed',
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    {
+        $validator = Validator::make(
+            $request->all(),
+            [
+                'email'    => 'required',
+                'password' => 'required',
+            ],
+            [
+                'email.required'    => 'Enter Your Email',
+                'password.required' => 'Enter Your Password',
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                'email.required'    => 'email is empty',
                'password.required' => 'password email is empty',
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                    'status'       => 'Success',
+                    'message'      => 'Login Success',
+                    'access_token' => $token,
+                    'token_type'   => 'Bearer',
+                ];
+                return response($response, 200);
+            } else {
+                $response = [
+                    'stutus'  => 'Failed',
+                    'message' => 'Password Incorrect',
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                    'stutus'  => 'failed',
                    'message' => 'incorrect password',
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        }
+        $user  = User::where('email', '=', $request->input('email'))->first();
+        if ($user) {
+            if (Hash::check($request->password, $user->password)) {
+                $token = $user->createToken('auth_token')->plainTextToken;
+                $response = [
+                    'status'       => 'Success',
+                    'message'      => 'Login Success',
+                    'access_token' => $token,
+                    'token_type'   => 'Bearer',
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                    'status'       => 'success',
                    'message'      => 'login success',
                    'access_token' => $token,
                    'token_type'   => 'bearer',
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+                $response = [
+                    'stutus'  => 'Failed',
+                    'message' => 'Password Incorrect',
+                ];
+                return response($response, 422);
+            }
+        } else {
+            $response = [
+                'status'  => 'Failed',
+                'message' => 'Wrong Email or Unregistered Email'
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                'status'  => 'failed',
                'message' => 'wrong email or unregistered email'
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        }
+    }
+
+    public function getprofile(Request $request)
+    {
+        return response()
+            ->json([
+                'Succes' => true,
+                auth()->user(),
+                'message' => 'Data Berhasil Ditampilkan'
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                'status' => 'success',
                auth()->user(),
                'message' => 'data berhasil ditampilkan'
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+                'email'    => 'required',
+                'password' => 'required',
+            ],
+            [
+                'email.required'    => 'Enter Your Email',
+                'password.required' => 'Enter Your Password',
+            ]
+        );
+        if ($validator->fails()) {
+            return response(['errors' => $validator->errors()->all()], 422);
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
            return response([
                'status' => `error`
                'message' => $validator->errors()->all()
            ], 422);
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    {
+        $validator = Validator::make(
+            $request->all(),
+            [
+                'name'     => 'required|string|max:255',
+                'password' => 'required|string|min:8',
+            ],
+            [
+                'name.required'     => 'Masukkan nama !',
+                'password.required' => 'Masukkan Password baru'
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                'name.required'     => 'name is empty',
                'password.required' => 'password is empty'
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+            ]
+        );
+        if ($validator->fails()) {
+            return response()->json([
+                'success' => false,
+                'message' => 'gagal',
+                'data'    => $validator->errors()
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                'status' => 'failed',
                'message' => 'update failed',
                'data'    => $validator->errors()
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+            $request->user()->fill([
+                'password' => Hash::make($request->password),
+            ])->save();
+            $user->save();
+            $valid = $user;
+            if ($valid) {
+                return response()->json([
+                    'success' => true,
+                    'id'      => $valid->id,
+                    'message' => 'sukses diperbarui',
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                    'status' => 'success',
                    'id'      => $valid->id,
                    'message' => 'update succeed',
```

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+                ], 200);
+            } else {
+                return response()->json([
+                    'success' => false,
+                    'message' => 'gagal',
```
Comment (chamisfum) [inline Student_HG_Laravel/app/Http/Controllers/AuthController.php]:
```suggestion
                    'status' => 'failed',
                    'message' => 'update failed',
```

**Review by chamisfum — CHANGES_REQUESTED** [review]
ada komen dikit ya @Hedo21 

> catatan : 
> 1. untuk json contract gunakan huruf kecil semua baik key atau valuenya.
> 2. untuk key gunakan `snake_case`
> 3. pastikan untuk menambahkan `status` dan `message` untuk setiap response yang ada.
> 4. untuk value boleh menggunakan bahasa inggris atau bahasa indonesia
> 5. untuk key gunakan bahasa inggris 

Oh iya untuk response success dan failed nya tolong sertakan curl nya ya mas

Terimakasih

**Comment by Hedo21** [comment]
oke mas @chamisfum cuman ku lanjutin jumat ya mas, ada dedlen tugas besar kuliah kamis mas, mau nyicil itu dulu

File: Student_HG_Laravel/app/Http/Controllers/AuthController.php (hunk 14 above)
Comment (chamisfum) [reply Student_HG_Laravel/app/Http/Controllers/AuthController.php]: Oke @Hedo21  lek bales disini aja mas ga perlu comment di tempat lain yaa

Comment (Hedo21) [reply Student_HG_Laravel/app/Http/Controllers/AuthController.php]: @chamisfum udah ku benerin mas

**Review by chamisfum — APPROVED** [review]
LGTM

**Review by chamisfum — COMMENTED** [review]
Need adding new variable `role`

Register request from FE
```json
{ 
   "name" : "string",
   "username" : "string" 
   "email": "string"
   "password": "string" 
   "role" : "string" // get from onboarding (dapat dari cache atau variable yang dikirim dari sesi on boarding)
}
```

Login request from FE
```json
{ 
   "email": "string"
   "password": "string" 
   "role" : "string" // get from onboarding (dapat dari cache atau variable yang dikirim dari sesi on boarding)
}
```

@Hedo21 

image

**Review by darmainchiro — APPROVED** [review]

**Comment by darmainchiro** [comment]
iyo neng heroguru-api-app

**Pull request merged**